[mono][mini] Interlocked.CompareExchange and Interlocked.Exchange intrinsics for small types and enums

lambdageek opened the pull request:
Only arm64, amd64 and WebAssembly. Interpreter, JIT and LLVM AOT. No x86 or arm32.

The complication for the 8- and 16-bit operands is that the .NET operand stack is 32-bit. so we have to zero- or sign-extend the inputs and output.

I ended up following the same approach as clang: zero-extend "expected" value and then sign-extend the return value of the CAS.  One nice consequence is that the LLVM support is essentially free.

---

Also this PR makes the intrinsic apply to types that have byte/int/long as the underlying type (bool, enums) not just when the type is literally in the signature

---

TODO:
- [x]  CompareExchange(u8/i8)
- [x] CompareExchange(u16/i16)
- [x] Exchange(u8/i8)
- [x] Exchange(u16/i16)
- [x] merge @kg jiterp branch into PR
- [x] interp+jiterp `Interlocked.Exchange(int)` intrinsics

---


Related to https://github.com/dotnet/runtime/issues/105335
Related to https://github.com/dotnet/runtime/issues/93488

lambdageek committed:
[mono][mini][arm64] Interlocked.CompareExchange for byte/sbyte

lambdageek committed:
rewrite cas_i1 with zext cmp and sign ext result

lambdageek committed:
[sample] fixup HelloWorld to work with full aot

lambdageek committed:
[mini][llvm] support OP_ATOMIC_CAS_U1

lambdageek committed:
[mini][amd64] support OP_ATOMIC_CAS_U1

lambdageek committed:
[mini][wasm] support OP_ATOMIC_CAS_U1 in LLVM AOT

lambdageek committed:
make intrinsic must-expand on arm64,amd64,wasm on mono

lambdageek committed:
[interp] MINT_MONO_CMPXCHG_U1

also add mono_atomic_cas_u8 utility function

lambdageek committed:
[mini] CompareExchange(i16/u16)

lambdageek committed:
[mono] must expand CompareExchange(i16/u16)

lambdageek committed:
[interp] Interlocked.CompareExchange(u16/i16)

lambdageek committed:
fix gcc intrinsics build

lambdageek commented:
Failures are related, although I'm not sure why yet - seemed like they were working when I only had u8 implemented. 

I wonder if we need to leave the fallbacks in place for partial gsharing and for gsharedvt

lambdageek committed:
fix amd64 supported ops

lambdageek commented:
Looks like a bunch of tests are failing due to `return Interlocked.CompareExchange(ref _isPendingInterrupt, false, true)` not working properly.  Need to investigate what's going on with bool

lambdageek committed:
[mini] zext unsigned CAS results for small types

lambdageek committed:
[interp] signed small CAS ops

lambdageek committed:
Interlocked.Exchange u1/i1/u2/i2 interp,llvm,amd64,arm64

lambdageek committed:
HACK: fix x64 crashes?

maybe we can't actually do a 1 or 2 byte move from RAX ?

lambdageek committed:
[amd64] fixup 8- and 16-bit cmpxchg encoding

kg commented on src/mono/mono/mini/interp/transform.c:
```diff
@@ -2459,13 +2459,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
-				*op = MINT_MONO_CMPXCHG_I8;
+		else if (!strcmp (tm, "CompareExchange") && csignature->param_count == 3) {
+			MonoType *t = mini_get_underlying_type (csignature->params[1]);
+			if (t->type == MONO_TYPE_I4 ||
```
do you really need to guard the switch with an if?

lambdageek replied:
probably not. I wasn't sure if I would make signed/unsigned opcodes or if I would emit a second instruction to fixup the result operand.  in the end it seemed better to have signed/unsigned opcodes.  I'll remove the "if" once I see a green x64 run

lambdageek committed:
fix operand order

lambdageek committed:
ok to always emit rex?

lambdageek commented on src/mono/mono/arch/amd64/amd64-codegen.h:
```diff
@@ -1216,9 +1216,25 @@ typedef union {
 #define amd64_cmpxchg_mem_reg_size(inst,mem,reg,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),0,0,(reg)); x86_cmpxchg_mem_reg((inst),(mem),((reg)&0x7)); amd64_codegen_post(inst); } while (0)
-#define amd64_cmpxchg_membase_reg_size(inst,basereg,disp,reg,size) do { amd64_codegen_pre(inst); amd64_emit_rex ((inst),(size),(reg),0,(basereg)); x86_cmpxchg_membase_reg((inst),((basereg)&0x7),(disp),((reg)&0x7)); amd64_codegen_post(inst); } while (0)
+
+#define amd64_cmpxchg_membase_reg_size(inst,basereg,disp,reg,size) \
+	do { \
+		amd64_codegen_pre(inst); \
+		if ((size) == 2) \
+			x86_prefix((inst), X86_OPERAND_PREFIX); \
+		amd64_emit_rex ((inst),(size),(reg),0,(basereg)); \
+		switch ((size)) { \
+		case 1: x86_byte((inst), 0x0f); x86_byte((inst), 0xb0); break; \
+		case 2: case 4: case 8: x86_byte((inst), 0x0f); x86_byte((inst), 0xb1); break; \
+		default: assert (0); \
+		}\
+		x86_membase_emit((inst),((reg)&0x7),((basereg)&0x7),(disp)); \
+		amd64_codegen_post(inst); \
+	} while (0)
```
@JulieLeeMSFT can someone from the JIT team please check that I'm doing x86-64 instruction encoding correctly here.

Things I'm unclear about:
1. can we really emit a REX prefix for 16-bit operands?  I'm not sure what in mono's codegen will guarantee that `reg` and `basereg` are the legacy registers if REX isn't allowed.
2. does this work if `reg` or `basereg` is one of the R8-R15 registers and size is 1? (apparently it worked when size was 4 or 8)

lambdageek replied:
Answering my own question: https://godbolt.org/z/jPEzErvGK

1. looks like `X86_OPERAND_PREFIX` and REX can co-exist.
2. seems to work if `reg` is one of the new registers and size is 1

kg committed:
If jiterpreter is engaged before the thread is fully initialized, just fail to allocate a table index and generate a warning. This shouldn't happen in prod anyway

Implement cmpxchg atomics natively in jiterpreter

Remove unnecessary jiterp cas helpers

Do cmpxchg result fixups as needed

Add runtime option for jiterpreter atomics
Implement atomic exchanges in the jiterpreter

kg commented:
I have a native jiterp implementation of all the cmpxchg opcodes (no C helper anymore) ready locally. Based on reading the spec we can just always use the native atomics opcode, and atomics have been available for a few years now so it should be safe to do that.

lambdageek commented on src/mono/mono/utils/atomic.h:
```diff
@@ -311,6 +327,18 @@ mono_atomic_store_ptr (volatile gpointer *dst, gpointer val)
+static inline guint16
+mono_atomic_cas_u16 (volatile guint16 *dest, guint16 exch, guint16 comp)
+{
+	return _InterlockedCompareExchange16 ((short volatile *)dest, (char)exch, (char)comp);
```
```suggestion
	return _InterlockedCompareExchange16 ((unsigned short volatile *)dest, exch, comp);
```

lambdageek committed:
Apply suggestions from code review

fixup u16 win32 atomic

lambdageek committed:
[interp] remove if guarding a switch

lambdageek commented:
@kg I added `Interlocked.Exchange(byte)` and `Interlocked.Exchange(ushort)` (and their signed counterparts) to the interp, LLVM AOT, arm64 and amd64 JIT, PTAL.

Probably ok to take the jiterp changes when you're ready.

---

For .NET 10 we should think about `Exchange` and `CompareExchange` for `object` and `i32/u32` for interp+jiterp (object is interesting because we'll need a GC write barrier).  Not sure if those are extra-risky to backport to 9 because of how common they probably are, or if they're less risky since there's no sext/zext required /cc @steveisok @lewing

matouskozak commented:
/azp run runtime-extra-platforms,runtime-llvm

lambdageek committed:
fix windows build

lambdageek commented:
/azp run runtime-extra-platforms,runtime-llvm

lambdageek commented:
One relevant failure on windows-x64

```
* Assertion at D:\a\_work\1\s\src\mono\mono\mini\mini-amd64.c:8072, condition `(code - cfg->native_code - offset) <= max_len' not met, function:mono_arch_output_basic_block, wrong maximal instruction length of instruction atomic_exchange_u2 (expected 12, got 13)
```

lambdageek committed:
[amd64] give u2 CMPXCHG and XCHG one more byte

for the 16-bit addressing prefix

lambdageek commented:
/azp run runtime-extra-platforms,runtime-llvm

kg reviewed (commented):
Interp parts look correct to me.

lewing commented:
@kg do you want to do the jiterp work here or in a follow up?

kg commented:
Jiterp implementation should be done already, I pushed it into the PR.

lewing commented:
> Jiterp implementation should be done already, I pushed it into the PR.

for all i/u 4 variants?

kg commented:
> > Jiterp implementation should be done already, I pushed it into the PR.
> 
> for all i/u 4 variants?

i/u 4 variants aren't in this PR, I think? They're on the todo list.

jkurdek commented on src/mono/sample/HelloWorld/HelloWorld.csproj:
```diff
@@ -44,6 +45,10 @@
   </Target>
 
+  <ItemGroup>
+      <ProjectReference Include="$(RepoTasksDir)AotCompilerTask\MonoAOTCompiler.csproj" ReferenceOutputAssembly="false" Condition="'$(RunAOTCompilation)' == 'true'" AdditionalProperties="Configuration=Debug"/>
```
For my education, why those changes were necessary? I thought we could already run the mono sample with FullAOT

lambdageek replied:
`./build.sh mono+libs+tasks` on OSX it doesn't bulid the AOT compiler task anymore because someone added a condition to `tasks.proj` to skip it on non-mobile platforms.  I think the long-term direction for tasks.proj is to get rid of it and add proper project references to all the projects that need tasks

jkurdek reviewed (approved):
arm changes look good to me

lambdageek commented:
> > Jiterp implementation should be done already, I pushed it into the PR.
> 
> for all i/u 4 variants?

I'm going to add i/u4 for the interpreter and jiterp today

lambdageek committed:
Interlocked.Exchange(int) for interp and jiterp

lambdageek commented on src/mono/mono/mini/interp/mintops.def:
```diff
@@ -709,7 +709,16 @@ OPDEF(MINT_LDPTR, "mono_ldptr", 3, 1, 0, MintOpShortInt)
 OPDEF(MINT_MONO_NEWOBJ, "mono_newobj", 3, 1, 0, MintOpClassToken)
 OPDEF(MINT_MONO_RETOBJ, "mono_retobj", 2, 0, 1, MintOpNoArgs)
 OPDEF(MINT_MONO_MEMORY_BARRIER, "mono_memory_barrier", 1, 0, 0, MintOpNoArgs)
+OPDEF(MINT_MONO_EXCHANGE_U1, "mono_interlocked.xchg.u1", 4, 1, 2, MintOpNoArgs)
+OPDEF(MINT_MONO_EXCHANGE_I1, "mono_interlocked.xchg.i1", 4, 1, 2, MintOpNoArgs)
+OPDEF(MINT_MONO_EXCHANGE_U2, "mono_interlocked.xchg.u2", 4, 1, 2, MintOpNoArgs)
+OPDEF(MINT_MONO_EXCHANGE_I2, "mono_interlocked.xchg.i2", 4, 1, 2, MintOpNoArgs)
+OPDEF(MINT_MONO_EXCHANGE_I4, "mono_interlocked.xchg.i4", 4, 1, 2, MintOpNoArgs)
 OPDEF(MINT_MONO_EXCHANGE_I8, "mono_interlocked.xchg.i8", 4, 1, 2, MintOpNoArgs)
+OPDEF(MINT_MONO_CMPXCHG_U1, "mono_interlocked.cmpxchg.u1", 5, 1, 3, MintOpNoArgs)
+OPDEF(MINT_MONO_CMPXCHG_I1, "mono_interlocked.cmpxchg.i1", 5, 1, 3, MintOpNoArgs)
+OPDEF(MINT_MONO_CMPXCHG_U2, "mono_interlocked.cmpxchg.u2", 5, 1, 3, MintOpNoArgs)
+OPDEF(MINT_MONO_CMPXCHG_I2, "mono_interlocked.cmpxchg.i2", 5, 1, 3, MintOpNoArgs)
```
@BrzVlad @kg instead of having a ton of different opcodes, we could have a single `MINT_MONO_EXCHANGE_INT` (and `MINT_MONO_CMPXCHG_INT`) opcode with a `isSigned << 4 | size` byte argument.  That would make the big opcode switch in interp.c a little smaller, but make the code a little more complicated.  Any opinions about which way is better?

lambdageek commented:
/backport to release/9.0

stephentoub commented:
Thank you, @lamdageek.
